src/platform/nm-platform-utils.c

```diff
@@ -404,7 +455,18 @@ nmp_utils_ethtool_set_link_settings (int ifindex,
 	edata.cmd = ETHTOOL_SSET;
 	if (autoneg) {
 		edata.autoneg = AUTONEG_ENABLE;
-		edata.advertising = edata.supported;
+		if (!speed && !duplex)
+			edata.advertising = edata.supported;
+		else {
+			unsigned long int mode;
+
+			mode = get_baset_mode (speed, duplex);
+
+			if (   !mode
+			    || !(edata.supported & mode))
+				return FALSE;
+			edata.advertising = edata.supported & ~(BASET_ALL_MODES & ~mode);
```

Comment:
edata.advertising is uint32 type. There is no need to pretend, mode could be an "unsigned long int". `get_baset_mode()` should return guint32. If you want, you can add a `G_STATIC_ASSERT()` that all `ADVERTISED_` flags fit into guint32, though not necessary. I think when doing bitwise operations, all operands should have the same integer type.

Reply:
```
edata.supported & ~(BASET_ALL_MODES & ~mode)
```
is the same as
```
(edata.supported & ~BASET_ALL_MODES) | (edata.supported & mode)
```
but the former seems hard to understand (to me).

Reply:
From ethtool.h:
"[...] Please do NOT define any SUPPORTED_* or ADVERTISED_ macro for bits > 31. [...]"
So, looks more than safe to use guint32 for the ADVERTISED_ flags.

As we check just above that (edata.supported & mode) is > 0, and mode is by construction just "one" mode (1 flipped bit), what about:
edata.advertising = (edata.supported & ~BASET_ALL_MODES) | mode  ?

Reply:
` (edata.supported & ~BASET_ALL_MODES) | mode`, yes even better. It's less about `mode` being just one bit (which is not immediately obvious), but more about mode being a subset of BASE_ALL_MODES (which seems obvious -- from the naming).